### PR TITLE
wip: receive frame from socket

### PR DESCRIPTION
**Pull request by nicboul**

### CODE REVIEW

**Review by si74 — APPROVED** [review]
👍

File: cmd/netvfy-agent/main.go [hunk 1]
```diff
@@ -371,20 +398,54 @@ func connController(ctx context.Context, cancel context.CancelFunc, ctrlInfo *co
...
+			dlog.Printf("Port: %s -- %s\n", gSwitch.info.Port, switchInfo.Port)
+			dlog.Printf("IPaddr: %s -- %s\n", gSwitch.info.IPaddr, switchInfo.IPaddr)
+			dlog.Printf("Netmask: %s -- %s\n", gSwitch.info.Netmask, switchInfo.Netmask)
 
 			gSwitch.info.Addr = switchInfo.Addr
 			gSwitch.info.Port = switchInfo.Port
 			gSwitch.info.IPaddr = switchInfo.IPaddr
+			gSwitch.info.Netmask = switchInfo.Netmask
+
+			cmd := exec.Command("ifconfig", utunName, gSwitch.info.IPaddr, gSwitch.info.IPaddr, "netmask", gSwitch.info.Netmask)
```
Comment (si74) [inline cmd/netvfy-agent/main.go:417]: can you include a comment //TODO: replace hardcoded portion with upstreamed tuntap functions?

File: cmd/netvfy-agent/main.go [hunk 2]
```diff
@@ -447,23 +512,27 @@ func connSwitch(ctx context.Context, cancel context.CancelFunc, config *tls.Conf
...
 				return
 				// the ticker is done
 			case <-done:
+				dlog.Printf("vswitch connection closing, ticker is done\n")
 				return
 			case <-ticker.C:
-				_, err := conn.Write(binBuf.Bytes())
+				_, err := vswitchConn.Write(keepaliveBuf.Bytes())
 				if err != nil {
+					dlog.Printf("got disconnected from the switch: %v\n", err)
```
Comment (si74) [inline cmd/netvfy-agent/main.go:535]: nit: // disconnected from switch

File: cmd/netvfy-agent/main.go [hunk 3]
```diff
@@ -473,7 +542,91 @@ func connSwitch(ctx context.Context, cancel context.CancelFunc, config *tls.Conf
 		}
 	}()
 
-out:
+	for {
+		select {
+		case <-done:
+			// ticker detected a disconnect form the controller
+			goto error
```
Comment (si74) [inline cmd/netvfy-agent/main.go:549]: hmm I've of mixed feelings about using goto in Go - sometimes makes code less readable in Go and is usually less of a goism/best practice imo. have you thought about using defer()'ed functions instead for some of the cleanup here and just return instead?

File: cmd/netvfy-agent/main.go [hunk 4]
```diff
@@ -110,8 +115,30 @@ var ilog *log.Logger
...
-
 var gNetConfPath string
+var utun *water.Interface
+var vswitchConn *tls.Conn
+
+var gMAC []byte
+
+const utunName = "utun7"
+
+func genMAC() []byte {
```
Comment (si74) [inline cmd/netvfy-agent/main.go:126]: +1

**Comment by si74** [comment]
And just one last comment - for now just for some readability, would be great if some code/functions could be split between files but all still in the package main

File: cmd/netvfy-agent/main.go [hunk 5]
```diff
@@ -417,22 +480,33 @@ func connSwitch(ctx context.Context, cancel context.CancelFunc, config *tls.Conf
...
 	var ticker *time.Ticker
-	var binBuf bytes.Buffer
+	var keepaliveBuf bytes.Buffer
 	var nvhdr *nvHdr
 	var state tls.ConnectionState
+	var offset int = 0
+
+	frameBuf := make([]byte, 2000)
+
+	defer func() {
```
Comment (si74) [inline cmd/netvfy-agent/main.go:490]: yay!

File: cmd/netvfy-agent/main.go [hunk 6]
```diff
@@ -417,22 +480,33 @@ func connSwitch(ctx context.Context, cancel context.CancelFunc, config *tls.Conf
 
 	var done chan bool
 	var ticker *time.Ticker
-	var binBuf bytes.Buffer
+	var keepaliveBuf bytes.Buffer
 	var nvhdr *nvHdr
 	var state tls.ConnectionState
+	var offset int = 0
+
+	frameBuf := make([]byte, 2000)
```
Comment (si74) [inline cmd/netvfy-agent/main.go:488]: oh interesting - I realized in my tuntap code i made this 1518 bytes - any reason why 2000 here?

**Review by si74 — APPROVED** [review]
👍

File: cmd/netvfy-agent/main.go (hunk 6 above)
Comment (nicboul) [reply cmd/netvfy-agent/main.go:488]: Mental lazyness, I'm sure it will be enough without thinking much about it :)

**Pull request merged**